Type error on registering plugin with custom options

### Expected behavior

The `BasePlugin` accepts a generic to be passed for the options of the plugin. Registering such plugins to the application cause type error. 

### Actual behavior

Registering such plugin is causing type error. 

```
error TS2345: Argument of type 'typeof MyPluign' is not assignable to parameter of type 'InstantiablePlugin<BasePlugin<PluginOptionsWithAppConfig>, PluginOptionsWithAppConfig>'.
  Types of parameters 'options' and 'args' are incompatible.
    Type 'PluginOptionsWithAppConfig' is missing the following properties from type 'PluginOptions': applicationUrl
```

Due to this error we can't compile and build the application. 

### Steps to reproduce

```js
interface MyOptions extends PluginOptionsWithAppConfig {
  applicationUrl: string;
}

class MyPlugin extends BasePlugin<MyOptions> {}

app.registerPlugin(MyPlugin)
```

### Which version(s) does this affect? (Environment, OS, etc...)
5.1.x

--- a/framework/src/application.ts
+++ b/framework/src/application.ts
@@ -33,7 +33,12 @@ import {
 	APP_EVENT_NETWORK_READY,
 } from './constants';
 
-import { getPluginExportPath, InstantiablePlugin, validatePluginSpec } from './plugins/base_plugin';
+import {
+	BasePlugin,
+	getPluginExportPath,
+	InstantiablePlugin,
+	validatePluginSpec,
+} from './plugins/base_plugin';
 import { systemDirs } from './system_dirs';
 import { Controller, InMemoryChannel } from './controller';
 import { applicationConfigSchema } from './schema';
@@ -172,8 +177,8 @@ export class Application {
 		return application;
 	}
 
-	public registerPlugin(
-		pluginKlass: InstantiablePlugin,
+	public registerPlugin<T extends BasePlugin>(
+		pluginKlass: InstantiablePlugin<T>,
 		options: PluginOptions = { loadAsChildProcess: false },
 	): void {
 		assert(pluginKlass, 'Plugin implementation is required');

--- a/framework/src/plugins/base_plugin.ts
+++ b/framework/src/plugins/base_plugin.ts
@@ -82,13 +82,12 @@ export interface PluginInfo {
 	readonly exportPath?: string;
 }
 
-export interface InstantiablePlugin<
-	T extends BasePlugin = BasePlugin,
-	U extends PluginOptionsWithAppConfig = PluginOptionsWithAppConfig
-> {
+type ExtractPluginOptions<P> = P extends BasePlugin<infer T> ? T : PluginOptionsWithAppConfig;
+
+export interface InstantiablePlugin<T extends BasePlugin = BasePlugin> {
 	alias: string;
 	info: PluginInfo;
-	new (args: U): T;
+	new (args: ExtractPluginOptions<T>): T;
 }
 
 const decodeTransactionToJSON = (

